File: app/src/main/java/com/zainco/library/dagger/daggerscopescasterio/AppComponent.java
package com.zainco.library.dagger.daggerscopescasterio;

import com.zainco.library.dagger.daggerscopescasterio.screens.profile.ProfileComponent;
import com.zainco.library.dagger.daggerscopescasterio.screens.profile.ProfileModule;
import com.zainco.library.dagger.daggerscopescasterio.screens.settings.SettingsComponent;
import com.zainco.library.dagger.daggerscopescasterio.screens.settings.SettingsModule;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {AppModule.class, AndroidModule.class})
public interface AppComponent {
    void inject(DemoApplication target);

    ProfileComponent plus(ProfileModule ProfileModule);

    SettingsComponent plus(SettingsModule settingsModule);
}

File: app/src/main/java/com/zainco/library/dagger/mitch/di/AppComponent.kt
package com.zainco.library.dagger.mitch.di

import android.app.Application
import com.zainco.library.dagger.mitch.BaseApplication
import com.zainco.library.dagger.mitch.SessionManager
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

/**
 * AndroidSupportInjectionModule: You always need to add it to the modules
 * ActivityBuildersModule: all the injected activities
 */
@Singleton
@Component(
    modules = [AndroidSupportInjectionModule::class,
        ActivityBuildersModule::class,
        AppModule::class,
        ViewModelFactoryModule::class]
)

interface AppComponent : AndroidInjector<BaseApplication> {
    //will be kept alive as long as application is alive
    fun sessionManager():SessionManager
    @Component.Builder
    interface Builder {
        //binds a particular object(BaseApplication class ) to the component at the time of its (respect to BaseApplication) construction
        @BindsInstance
        fun application(baseApplication: Application): Builder

        fun build(): AppComponent
    }
}